ocaml/xapi/xapi_host.ml

```diff
@@ -3055,18 +3060,24 @@ let try_restart_device_models_for_recommended_guidances ~__context ~host =
       None
 
 let apply_recommended_guidances ~__context ~self =
+  (* This function runs on master host *)
+  Helpers.assert_we_are_master ~__context ;
   try
     let open Updateinfo in
     Db.Host.get_recommended_guidances ~__context ~self |> function
     | [] ->
         try_restart_device_models_for_recommended_guidances ~__context
           ~host:self
     | [`reboot_host] ->
-        reboot ~__context ~host:self
+        Helpers.call_api_functions ~__context (fun rpc session_id ->
+            Client.Client.Host.reboot ~rpc ~session_id ~host:self
+        )
     | [`restart_toolstack] ->
         try_restart_device_models_for_recommended_guidances ~__context
           ~host:self ;
-        restart_agent ~__context ~host:self
+        Helpers.call_api_functions ~__context (fun rpc session_id ->
+            Client.Client.Host.restart_agent ~rpc ~session_id ~host:self
```

Comment:
I expect this to close the connection, because xapi will be restarted, is this handled in a graceful way here, does it appear as an error instead?

Reply:
As `restart_agent` is called with a new thread, I think there will be no error calling `restart_agent`. While in the new thread requesting a restart of xapi or the host will make the connection closed and thus error will arise, I think this is inevitable, or would you please share what is the "graceful way" to handle it? Thank you!